packages/teleterm/src/ui/Navigator/ExpanderConnections/useExpanderConnections.ts

```diff
@@ -18,81 +18,108 @@ import * as Icons from 'design/Icon';
-      uri: '/nothing/402',
-      Icon: Icons.Keypair,
-      items: [],
-      group: false,
-      status: 'disconnected',
+function getWorkspaceItems(ctx: AppContext): ConnectionItem[] {
+  return [
+    ...ctx.serviceWorkspace
+      .get()
+      .recentDocuments.map(document => {
+        if (document.kind === 'doc.terminal_tsh_node') {
+          return { ...document, status: 'disconnected' as const };
+        } else {
+          return document;
+        }
+      })
+      .map(document => createConnectionItem(ctx, document)),
+  ].sort(item => {
+    if (item.status === 'connected') {
+      return -1;
+    }
+    return 1;
+  });
+}
+
+function handleItemOpen(ctx: AppContext, item: ConnectionItem) {
+  if (!ctx.serviceDocs.getDocuments().find(d => d.uri === item.uri)) {
+    ctx.serviceDocs.add(item.document);
+  }
+  ctx.serviceDocs.open(item.uri);
+}
+
+function handleItemRemove(ctx: AppContext, item: ConnectionItem) {
+  const recentDocuments = ctx.serviceWorkspace
+    .get()
+    .recentDocuments.filter(d => d.uri !== item.uri);
+  ctx.serviceWorkspace.update({ recentDocuments });
+}
+
+function createConnectionItem(
+  ctx: AppContext,
+  document: Document
+): ConnectionItem {
+  function getStatus() {
+    switch (document.kind) {
+      case 'doc.terminal_tsh_node':
+        return (ctx.serviceDocs.getDocument(document.uri) as DocumentTshNode)
+          ?.status === 'connected'
+          ? 'connected'
+          : 'disconnected';
+      case 'doc.gateway':
+        return ctx.serviceClusters.findGateway(document.uri)
+          ? 'connected'
+          : 'disconnected';
+    }
+  }
+
+  return {
+    get status() {
+      return getStatus();
     },
-    {
-      title: 'pearl@node1',
-      kind: 'nav.connection-tsh',
-      uri: '/nothing/2cf',
-      Icon: Icons.Keypair,
-      items: [],
-      group: false,
-      status: 'disconnected',
+    get uri() {
+      return document.uri;
     },
-    {
-      title: 'lulu@node1',
-      kind: 'nav.connection-tsh',
-      uri: '/nothing/23b',
-      Icon: Icons.Keypair,
-      items: [],
-      group: false,
-      status: 'disconnected',
+    get title() {
+      return document.title;
     },
-  ];
-
-  const tshItems: ConnectionItem[] = tshDocs.map(d => ({
-    title: d.title,
-    status: 'connected',
-    kind: 'nav.connection-tsh',
-    uri: d.uri,
     Icon: Icons.Keypair,
     items: [],
     group: false,
-  }));
-
-  const gwItems: ConnectionItem[] = gateways.map(g => ({
-    title: g.resourceName,
-    status: 'connected',
-    Icon: Icons.Keypair,
-    uri: g.uri,
-    kind: 'nav.connection-gw',
-    items: [],
-    group: false,
-  }));
-
-  return [...tshItems, ...gwItems, ...demoHistoryItems];
+    document,
+  };
 }
 
 export interface ConnectionItem extends types.NavItem {
-  status: 'connected' | 'disconnected';
-  kind: 'nav.connection-tsh' | 'nav.connection-gw';
+  readonly status: 'connected' | 'disconnected';
+  document: Document;
```

Comment:
IMO we don't need this additional kind (as we have a document type)

Reply:
This would introduce a dependency on the `Document` on unnecessary levels. I would take what i need from the document to initiate the items and then let go of the document after that (if possible). This way rendering of an icon for the item will not require any Document references.

Reply:
This would introduce a dependency on the `Document` on unnecessary levels. I would take what i need from the document to initiate the items and then let go of the document after that (if possible). This way rendering of an icon for the item will not require any Document references.

Reply:
Agree, no need to pass the `Document` down